### ISSUE TITLE
Slight change in tag datastructure

### ISSUE THREAD

**Issue by JMurphyWeb** [emfoundation/ce100-app#350]
Old: 
```js
[ { id: 0,
    name: 'MEMBER TYPE ',
    enabled: true,
    tags:
     [ [Object],
       [Object],
       [Object] ],
    type: 'Member types' },
  { id: 1,
    name: 'SECTOR/DEPARTMENT ',
    enabled: true,
    tags:
     [ [Object],
     ...
 }
]
```

New:
```js
[
  {
    id_cat: theIdeOfTheCategory,
    name_cat: theNameOfTheCategory,
    tags: [
       {id_tag: theIdOfTheTag, name_tag: theNameOfTheTag},...
    ]
  },
 ...
]

```

**Comment by JMurphyWeb** [comment]
this is how we render a tag:

```
<label class="tag__title" for="{{@index}}">{{name}}</label>
<input  type="checkbox"
        class="tag__input"
        id="{{@index}}">

<div class="form__subsection">
    {{#each tags}}
        {{#if ../view.challenges}}
          <a class="text-decoration" href="/challenges?tags={{id}}">
        {{else if ../view.orgs}}
          <a class="text-decoration" href="/orgs?tags={{id}}">
        {{/if}}
            <span class="tag__label" value=[{{id}}]>{{name}}</span>
          </a>
    {{/each}}
</div>

```

**Comment by JMurphyWeb** [comment]
@SimonLab I think we want more specific names: 
`cat_id, cat_name, tag_id, tag_name`

**Comment by JMurphyWeb** [comment]
@mk4111 @SimonLab  Presumably we will still send the id of a single tag up as `[cat_id, tag_id]` so that we know which category the selected child tag sits in?

**Comment by JMurphyWeb** [comment]
@mk4111 @SimonLab One other question: 
If 1 tag sits in 2 categories, say an admin wants to slightly change the name of one of the tags, that will be quite hard in the way that we have done it. Have we thought about this scenario?

**Comment by mk4111** [comment]
@JMurphyWeb, I like it, I think we need to push up both id_cat and id_tag because, with the way that we are currently rendering the 'select-tags-view', we toggle the 'parent-tag'/category, to show all the 'child-tags'/tags associated with that category. 

I also agree that there are a few scenarios that I don't think we've thought through thoroughly enough; would this be the right place to discuss them?
(1)
> If 1 tag sits in 2 categories, say an admin wants to slightly change the name of one of the tags, that will be quite hard in the way that we have done it. Have we thought about this scenario?

We will have to find a way to _create a new tag_, go through the joint table and find the relationship between the tag they want to rename and categories to which they are attached, remove the relationship and replace it with the new id and category relationship I think.... 

(2) When we edit, for example, a challenge - we highlight the tags that have already been displayed; similar to your question above, how do we know which they are associated with? I think we would need to store in the joint table, the id_tag, id_challenge but also the id_cat too.

**Comment by JMurphyWeb** [comment]
@SimonLab was saying that if a tag is put into more than one category, then it represents the same tag. If the admin changes the name of it, the name should change in each category that the tag appears.

**Comment by mk4111** [comment]
This is what I meant...

![screenshot from 2016-11-18 09-32-17](https://cloud.githubusercontent.com/assets/13470325/20425317/040e0be4-ad72-11e6-9555-ff13f2ac262c.png)

**Comment by JMurphyWeb** [comment]
@mk4111 is that to do with this comment?
> 
@JMurphyWeb, I like it, I think we need to push up both id_cat and id_tag because, with the way that we are currently rendering the 'select-tags-view', we toggle the 'parent-tag'/category, to show all the 'child-tags'/tags associated with that category.

Can you expand at all, I dont quite follow..

**Comment by mk4111** [comment]
@JMurphyWeb - yes it was to do with that comment :)

**Comment by mk4111** [comment]
@JMurphyWeb - I agree with SimonLab too, but I wonder what would happen in the scenario you just described... It works, but only to a certain extent, not if you want to edit the existing tag to suit just one of the categories; I guess in this case they would just need to create a new tag!

**Comment by JMurphyWeb** [comment]
can you explain why it is important we send the category id up to the handlers?

**Comment by mk4111** [comment]
@JMurphyWeb - maybe not id_cat, but name_cat will be needed to display it on [this view](https://github.com/emfoundation/ce100-app/issues/350#issuecomment-261487608) right?

**Comment by mk4111** [comment]
@JMurphyWeb and @SimonLab, although the tags are unique in their own dataset, they won't be unique once they become shared between categories, so surely we need to find a way to keep note of that relationship too?

**Comment by JMurphyWeb** [comment]
@mk4111 I think Simon is suggesting that if a tag exists in more than one category, if it is 'selected', then it will be 'selected' in all of those categories

**Comment by JMurphyWeb** [comment]
I think this snippet shows what it would look like, let me know if something doesnt look right:
```html
<label class="tag__title" for="{{@index}}">{{name_cat}}</label>
<input  type="checkbox"
        class="tag__input"
        id="{{@index}}">

<div class="form__subsection">
    {{#each tags}}
        {{#if ../view.challenges}}
          <a class="text-decoration" href="/challenges?tags={{id_tag}}">
        {{else if ../view.orgs}}
          <a class="text-decoration" href="/orgs?tags={{id_tag}}">
        {{/if}}
            <span class="tag__label" value="{{id_tag}}">{{name_tag}}</span>
          </a>
    {{/each}}
</div>

```

**Comment by mk4111** [comment]
@JMurphyWeb - if it's selected in all the categories, is there a point in having the tags-categories joint table...? :/ On edit view, the tags will show up as selected in multiple categories, which I think could be very confusing for the user...

**Comment by JMurphyWeb** [comment]
@mk4111 Yeah I agree, maybe there is a way around that, or maybe we can wait for a user to give that sort of feedback?

**Comment by SimonLab** [comment]
Instead of building an array for the new structure we can build easily a new object: 
```
{
  idCat: {
    nameCat: ---,
   tags: {
      idTag: {nameTag: ---}, ...
    }
  },
...
}
```
where idCat will be the id of the category.
The example is just to show the structure of the object and we can change the name of the properties and add some more inside the objects

So it can looks something like this:

```
{
  1: {
    idCat: 1,
    nameCat: 'Category 1',
    tags: {
      1: {idTag: 1, nameTag: 'tag 1'},
      2: {idTag: 2, nameTag: 'tag 2'}
    }
  },
...
}
```

**Comment by JMurphyWeb** [comment]
@SimonLab why is this better? 

We cannot loop through it as easily as we currently do in handlebars...

**Comment by JMurphyWeb** [comment]
Another thought, we have a load of helper functions already built that should play nicely if we use old structure.

**Comment by SimonLab** [comment]
This structure is much simpler to build than the other one with javascript. At the moment I haven't find an easy way to build the other structure but if you find one we can keep the other structure.
you can loop on object with handlebars too:
```
{{#each object}}
  {{@key}}: {{this}}
{{/each}}
```
The helpers will still have to change a bit as the structure will still be a bit different than the original one.

**Comment by mk4111** [comment]
Because the ids are the keys, if we wanted to access their values, would this structure impede us from doing so?

**Comment by SimonLab** [comment]
The structure defined is the [previous comment](https://github.com/emfoundation/ce100-app/issues/350#issuecomment-261502493) is now returned by the function getTags implemented on the PR #349
